### PR TITLE
Publish Arm's internal documentation

### PR DESCRIPTION
**Pull request by michalpasztamobica**
This is the first part of the exercise aiming at establishing the approach, files locations, agreeing on using `rst` format, etc.

The highlights of this commits are:
1) New documents added in `.rst` format in the main `docs/` directory. Being aware that most documentation so far was only available in the `.nroff` format we see some befits of choosing `rst` for new docs.
* `rst` to html on the fly, so all our documents will be viewable online
* `rst` files can also be used for the wiki (unless we decide to move wiki to main repository)
* current build tools allow us to convert `rst` to html easily using Sphinx and generate neat html files (see [html_docs_flang.zip](https://github.com/flang-compiler/flang/files/5401472/html_docs_flang.zip)), which we could set up somewhere - this is why I would prefer `rst` over `md` format.
2) `docs/ReleaseNotes.rst` is removed - it has never been used
3) Glossary is introduced based on ARM's glossary with some more phrases added.

Please review the code thinking wide - if this structure is approved we'll add more articles and documents in the near future.

@kiranchandramohan , @RichBarton-Arm , thanks for internal reviews and sharing Arm's internal documentation!

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
ping. Is this OK with everyone? This is only a documentation change.

File: tools/flang1/utils/n2rst/CMakeLists.txt
Comment (bryanpkc) [inline tools/flang1/utils/n2rst/CMakeLists.txt:20]: Don't we need to create the directory before writing files to it?

Comment (bryanpkc) [reply tools/flang1/utils/n2rst/CMakeLists.txt:20]: `tools/flang1/flang1exe/CMakeLists.txt` also creates the directory, but only when `FLANG_INCLUDE_DOCS`  is true. That may result in situation where the `n2rst` build needs the directory but cannot find it.

**Review by gklimowicz — APPROVED** [review]
These new documents got built for me on OpenPower.

File: tools/flang1/utils/n2rst/CMakeLists.txt
Comment (michalpasztamobica) [reply tools/flang1/utils/n2rst/CMakeLists.txt]: This directory is created by two other `CMakeLists.txt`s already, but I guess you're right - it doesn't hurt if we keep this command in here, as it generates no errors or warnings and will keep this script working if other `CMakeLists.txt`s alter.

Comment (bryanpkc) [reply tools/flang1/utils/n2rst/CMakeLists.txt]: Thanks.

**Review by bryanpkc — APPROVED** [review]
LGTM.

**Comment by michalpasztamobica** [comment]
The reviews were still pending, so I allowed myself to expand this PR instead of creating a new one, as I already got green light from @RichBarton-Arm  to publish further docs. I hope this is OK?
@gklimowicz , @bryanpkc , would you please look at the other commit I just pushed and review that part?
@shivaramaarao , @RichBarton-Arm, your reviews will be very welcome for the whole PR.
@kiranchandramohan , once the approvals are there, I'll squash the two commits into one.

**Comment by gklimowicz** [comment]
> The reviews were still pending, so I allowed myself to expand this PR instead of creating a new one, as I already got green light from @RichBarton-Arm to publish further docs. I hope this is OK?

Some of us had already reviewed it. In my case, I would have preferred to have a new PR for the additional changes.

**Review by gklimowicz — APPROVED** [review]
LGTM

File: docs/glossary.rst
Comment (bryanpkc) [inline docs/glossary.rst:16]: Nit: There is no need for a comma after a parenthesized clause, and I would put the main definition first, e.g.
```
short for "backend" (used heavily in code and comments)
```

Comment (bryanpkc) [reply docs/glossary.rst:16]: The same comment applies to all rows below that have parenthetical clauses.

File: docs/debugging_flang.rst
Comment (bryanpkc) [inline docs/debugging_flang.rst:73]: Would you consider styling all options in this table consistently as preformatted code? For example <code>\`\`filename\`\`</code>, <code>\`\`opt\`\`</code>, <code>\`\`x\`\`</code>, <code>\`\`qq\`\`</code>, etc.

File: docs/debugging_flang.rst
Comment (bryanpkc) [inline docs/debugging_flang.rst:101]: Should this explain that `y` flags turn off the corresponding `x` flags?

**Comment by michalpasztamobica** [comment]
I am sorry @gklimowicz , next time in a situation like this I'll set up a new PR. Thanks for a quick review!
@bryanpkc , I hope I addressed all your remarks? I put them in a separate commit, which I will squash once I get all reviews.

@shivaramaarao , @RichBarton-Arm , any remarks to this PR?

**Review by bryanpkc — APPROVED** [review]
Thanks, LGTM.

**Review by shivaramaarao — APPROVED** [review]
Looks fine for me.

**Review by RichBarton-Arm — APPROVED** [review]

**Comment by RichBarton-Arm** [comment]
Thanks @michalpasztamobica

**Comment by bryanpkc** [comment]
@RichBarton-Arm Please rebase and merge; do not create merge commit in the git history.

**Comment by RichBarton-Arm** [comment]
@bryanpkc - arg, sorry. I should have left it to Kiran to push the button.

Shall we try to fix it up?

**Comment by bryanpkc** [comment]
@kiranchandramohan @gklimowicz @shivaramaarao What do you think? It's probably okay to fix the branch right?

Is there a setting in GitHub to prevent non-fast-forward merges?

**Comment by michalpasztamobica** [comment]
Please, feel free to squash the three commits into one, if editing the master branch.

**Comment by kiranchandramohan** [comment]
I am OK with fixing. We can discuss it tomorrow during the classic flang call.

**Pull request merged**
